File: backend/app/schemas/simulation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class SimulationRunRequest(BaseModel):
    tick_duration_ms: int = 2000
    max_ticks: int = 60


class SimulationStatusResponse(BaseModel):
    scenario_id: UUID
    status: str
    current_tick: int
    max_ticks: int
    agents_active: int
    interactions_count: int


class SimulationResultResponse(BaseModel):
    id: UUID
    scenario_id: UUID
    p_impact: float
    buy_ratio: float
    sell_ratio: float
    hold_ratio: float
    sentiment_timeline: list
    narrative_summary: str | None
    emergent_narratives: list
    top_influencers: list
    raw_metrics: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class HeatmapDataPoint(BaseModel):
    persona_type: str
    tick: int
    avg_sentiment: float
    action_counts: dict


class HeatmapResponse(BaseModel):
    data: list[HeatmapDataPoint]
    personas: list[str]
    ticks: int


class PriceImpactPoint(BaseModel):
    tick: int
    p_impact: float
    by_persona: dict


class PriceImpactResponse(BaseModel):
    data: list[PriceImpactPoint]


class NarrativeResponse(BaseModel):
    summary: str
    emergent_narratives: list[dict]
    top_influencers: list[dict]

File: frontend/types/simulation.ts
export interface SimulationStatus {
  scenario_id: string;
  status: string;
  current_tick: number;
  max_ticks: number;
  agents_active: number;
  interactions_count: number;
}

export interface SimulationResult {
  id: string;
  scenario_id: string;
  p_impact: number;
  buy_ratio: number;
  sell_ratio: number;
  hold_ratio: number;
  sentiment_timeline: TickMetrics[];
  narrative_summary: string | null;
  emergent_narratives: EmergentNarrative[];
  top_influencers: TopInfluencer[];
  raw_metrics: Record<string, unknown>;
  created_at: string;
}

export interface TickMetrics {
  tick: number;
  avg_sentiment: number;
  by_persona: Record<string, number>;
  p_impact: number;
  stance_ratios: {
    buy_ratio: number;
    sell_ratio: number;
    hold_ratio: number;
  };
  interaction_count: number;
}

export interface TickUpdate {
  type: "tick_update";
  tick: number;
  interactions: TickInteraction[];
  metrics: TickMetrics;
}

export interface TickInteraction {
  id: string;
  agent_name: string;
  persona_type: string;
  platform: string;
  content: string;
  sentiment: number;
  stance: "BUY" | "SELL" | "HOLD";
}

export interface EmergentNarrative {
  type: string;
  description: string;
  severity: string;
}

export interface TopInfluencer {
  name: string;
  persona: string;
  count: number;
}
